### DISCUSSION TITLE
Notable fork submissions

### DISCUSSION THREAD

**Discussion by svlandeg** [karpathy/autoresearch#225] [Show and tell]
For easier reviewing, I thought I'd collect some of the submissions here.

- [ ] [sanbuphy/autoresearch-cn](https://github.com/sanbuphy/autoresearch-cn): Chinese community ([#4](https://github.com/karpathy/autoresearch/issues/4), [#240](https://github.com/karpathy/autoresearch/discussions/240))
- [ ] [sebasmos/autoresearch-dj](https://github.com/sebasmos/autoresearch-dojo): DOJO adversarial testing on Apple Silicon ([75](https://github.com/karpathy/autoresearch/pull/75))
- [ ] [Aum08Desai/test-time-rl-discover-autoresearch](https://github.com/Aum08Desai/test-time-rl-discover-autoresearch): Test Time RL Training on Autoresearch  ([94](https://github.com/karpathy/autoresearch/pull/94), [229](https://github.com/karpathy/autoresearch/pull/229))
- [ ] [eimenhmdt/autoresearcher](https://github.com/eimenhmdt/autoresearcher): Research enhancements)  ([#123](https://github.com/karpathy/autoresearch/pull/123))
- [ ] [hwchase17/autoresearch-agents](https://github.com/hwchase17/autoresearch-agents): Multi-agent improvements)  ([#123](https://github.com/karpathy/autoresearch/pull/123))
- [ ] [ncdrone/autoresearch-ANE](https://github.com/ncdrone/autoresearch-ANE): Apple Neural Engine  ([#123](https://github.com/karpathy/autoresearch/pull/123))
- [ ] [JKeller45/autoresearch-rocm](https://github.com/JKeller45/autoresearch-rocm): AMD ROCm ([#150](https://github.com/karpathy/autoresearch/discussions/150))
- [ ] [mmonad/autoresearch-dgx-spark](https://github.com/mmonad/autoresearch-dgx-spark): NVIDIA DGX Spark / GB10 ([#207](https://github.com/karpathy/autoresearch/pull/207))
- [ ] [parthwhy/autoresearch-lite](https://github.com/parthwhy/autoresearch-lite) :  run on free T4 GPUs (Google Colab / Kaggle) ([#208](https://github.com/karpathy/autoresearch/issues/208))
- [ ] [ajzhanghk/autoresearch-glm](https://github.com/ajzhanghk/autoresearch-glm): Autonomous feature discovery for tabular GLM models  ([#209](https://github.com/karpathy/autoresearch/issues/209))
- [ ] [Entrpi/autoresearch-everywhere](https://github.com/Entrpi/autoresearch-everywhere): Optimized cross-platform expansion of autoresearch
- [ ] [matt-k-wong/autoresearch_mlx_mkw](https://github.com/matt-k-wong/autoresearch_mlx_mkw): MacOS / Optimized for 16GB RAM with Linear Attention & Muon ([#227](https://github.com/karpathy/autoresearch/issues/227))
- [ ] [autoresearch-rustbrain](https://github.com/m0at/autoresearch-rustbrain): Rust + CUDA reimplementation ([#231](https://github.com/karpathy/autoresearch/issues/231))
- [ ] [morozow/autoresearch-stdio-bus](https://github.com/morozow/autoresearch-stdio-bus): Quantum-Born Consciousness ([#234](https://github.com/karpathy/autoresearch/discussions/234))
- [ ] [Venkateshwar-PortoAI/autoresearch-mac-mini](https://github.com/Venkateshwar-PortoAI/autoresearch-mac-mini): Mac Mini, PyTorch MPS/CPU, no CUDA required ([#252](https://github.com/karpathy/autoresearch/pull/252))
- [ ] [lucasgelfond/autoresearch-webgpu](https://github.com/lucasgelfond/autoresearch-webgpu): WebGPU / browser ([#253](https://github.com/karpathy/autoresearch/pull/253))

**Comment by svlandeg** [comment]
If you'd like to add yours, feel free to open a new thread in Show and Tell, explain it there in detail, and then link it here with a one-liner as above!

**Reply by ar0cket1** [reply]
[https://github.com/Aum08Desai/test-time-rl-discover-autoresearch]: Test Time RL Training on Autoresearch (https://github.com/karpathy/autoresearch/discussions/229) (https://github.com/karpathy/autoresearch/pull/94)

**Reply by supratikpm** [reply]
[supratikpm/gemini-autoresearch](https://github.com/supratikpm/gemini-autoresearch) — 
Generalised autoresearch skill for Gemini CLI and Antigravity IDE. 
Uses Google Search grounding as a live verification source inside the loop. 
Full write-up: #413

**Comment by matt-langston** [comment]
Hi @svlandeg,

Would you delete mine from the list please? Second from the top "NVIDIA DGX Spark Bundle (2 x GB10) (https://github.com/karpathy/autoresearch/discussions/67)". Thank you!

**Reply by Muminur** [reply]
https://github.com/Muminur/autoresearch-skill-Andrej-Karpathy

**Comment by Entrpi** [comment]
<img width="1370" height="429" alt="autoresearch-everywhere" src="https://github.com/user-attachments/assets/5bd47d29-4544-4a06-879c-080067473fbb" />

# https://github.com/Entrpi/autoresearch-everywhere

Headline feature: One button calibration for your hardware. It is meant to run on anything and well.

This is the most actively developed fork of all, as I'd already been working on a similar project when autoresearch released and was able to port ideas into it. It is already cross-platform and tested on Apple M5 and DGX Spark (working with FlashAttention 4 too), with future ROCm support planned. There are many optimizations to the engine, and I've also built it to do autoresearch on kernels. I also developed [EEmicroGPT](https://github.com/Entrpi/eemicrogpt) (the fastest implementation of microgpt by ~10x and ~20,000x faster than the CPython upstream) and will be using that work to add highly optimized pure-CPU training as well soon.

**Comment by EdwardOptimization** [comment]
Hi @svlandeg,
  Here is my idea and some codes.
  [EdwardOptimization/Bilevel-Autoresearch](https://github.com/EdwardOptimization/Bilevel-Autoresearch): 
Adds a second optimization loop — the outer loop analyzes inner loop traces and autonomously generates new pipeline stages (Python code via DeepSeek). Detailed write-up in [#375](https://github.com/karpathy/autoresearch/discussions/375).
  
  The key Idea is "If autoresearch is a research, then autoresearch can research itself."

**Comment by ediestel** [comment]
https://github.com/ediestel/picochat-mlx.  Nanochat in MLX Based on Awni Hannun - picochat; with modifications

**Comment by supratikpm** [comment]
[supratikpm/gemini-autoresearch](https://github.com/supratikpm/gemini-autoresearch) — 
Generalised autoresearch skill for Gemini CLI and Antigravity IDE. 
Uses Google Search grounding as a live verification source inside the loop. 
Full write-up: #413

**Comment by mazar** [comment]
A working fork for NVIDIA DGX Spark: https://github.com/mazar/autoresearch-spark

```
Vocab size: 8,192
Model config: {'sequence_len': 2048, 'vocab_size': 8192, 'n_layer': 6, 'n_head': 3, 'n_kv_head': 3, 'n_embd': 384, 'window_pattern': 'L'}
Parameter counts:
  wte                     : 3,145,728
  value_embeds            : 9,437,184
  lm_head                 : 3,145,728
  transformer_matrices    : 10,617,120
  scalars                 : 12
  total                   : 26,345,772
Estimated FLOPs per token: 1.392002e+08
Scaling AdamW LRs by 1/sqrt(384/768) = 1.414214
Time budget: 300s
Gradient accumulation steps: 1
W0429 22:54:05.451000 24942 .venv/lib/python3.10/site-packages/torch/_inductor/utils.py:1731] [0/0] Not enough SMs to use max_autotune_gemm mode
step 00631 (100.0%) | loss: 3.224842 | lrm: 0.00 | dt: 469ms | tok/sec: 279,497 | mfu: 15.6% | epoch: 1 | remaining: 0s     
---
val_bpb:          1.149530
training_seconds: 300.4
total_seconds:    362.1
peak_vram_mb:     13706.3
mfu_percent:      15.11
total_tokens_M:   82.8
num_steps:        632
num_params_M:     26.3
depth:            6
```

**Comment by tal-giladi** [comment]
**CPU-only fork, plus a course built on top of it**

[tal-giladi/autoresearch](https://github.com/tal-giladi/autoresearch) — runs the same `train.py` on a laptop CPU inside docker, no GPU and no NVIDIA runtime. What changed (every deviation is tagged `# CHANGED` with a reason):

- flash-attn-3 → `F.scaled_dot_product_attention` plus an explicit causal+sliding-window mask, verified equal to the CUDA path to 1e-6
- fp32 instead of bf16 on CPU — measured bf16 at **0.78x** of fp32 here, since most x86 has no bf16 matmul instruction
- `torch.compile` off (no toolchain in the slim image, and the warm-up would eat the budget), `torch.cuda.*` guarded, pinned-host staging buffer dropped
- toy-scale defaults and an `AR_*` env override for every knob; with none set and a GPU present the defaults are exactly upstream's
- the first-N untimed steps drop from 10 to 2 — with compile off those were minutes of free, untimed training
- the `\r` progress line gets a newline when stdout is not a TTY, otherwise docker's log driver buffers the whole run and shows nothing until it exits
- saves a checkpoint and adds `sample.py`, so you can actually read what the model learned

`prepare.py` keeps the fixed metric; `EVAL_TOKENS` and `MAX_SEQ_LEN` are lowered fork-wide and declared loudly, so runs are internally comparable but **not** comparable to upstream's numbers.

Measured on 16 threads (~30 GFLOP/s fp32 peak): ~300 tok/s, 11.5M params, `val_bpb` 2.40 after a 10-minute run, MFU ~4.5%. It is a toy — 0.016 tokens per parameter, about 1000x short of compute-optimal — and the README says so.

[tal-giladi/gpt-from-scratch-course](https://github.com/tal-giladi/gpt-from-scratch-course) — a 7-module, 20-lesson course that teaches this `train.py` line by line, for engineers with no ML background. Every exercise is **graded against the real repo** inside that same CPU image (`bash lab/lab.sh check NN`): you write the packer, causal windowed attention, rotary, RMS norm, the ReLU² MLP, the parameter-count formula, sampling, an AdamW step, Muon's momentum, the LR schedule, gradient accumulation, `bits_per_byte`, `estimate_flops`, the window mask and a run-log parser, and each is compared against the actual implementation. The capstone makes you run baseline / repeat / variant, measure your own noise floor, and it fails you if your conclusion is inside it.

Rendered: https://tal-giladi.github.io/gpt-from-scratch-course/

**Comment by crivayne** [comment]
[crivayne/autoresearch-xv](https://github.com/crivayne/autoresearch-xv): vendor-agnostic single source (NVIDIA+AMD from one train.py) for measuring the cross-vendor efficiency gap; RDNA4/ROCm with torch.compile+TunableOp on native Linux ([Show and tell](https://github.com/karpathy/autoresearch/discussions/641))
